influxql/ast.go

```diff
@@ -876,6 +906,48 @@ func (s *SelectStatement) Validate(tr targetRequirement) error {
 		}
 	}
 
+	if err := s.validateDerivative(); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (s *SelectStatement) validateDerivative() error {
+	if !s.HasDerivative() {
+		return nil
+	}
+
+	// If a derivative is requested, it must be the only field in the query. We don't support
+	// multiple fields in combination w/ derivaties yet.
+	if len(s.Fields) != 1 {
+		return fmt.Errorf("derivative cannot be used with other fields")
+	}
+
+	aggr := s.FunctionCalls()
+	if len(aggr) != 1 {
+		return fmt.Errorf("derivative cannot be used with other fields")
+	}
+
+	// Derivative requires two arguments
+	derivativeCall := aggr[0]
+	if len(derivativeCall.Args) != 2 {
+		return fmt.Errorf("derivative requires a field and duration argument")
```

Comment:
The duration can be different than the group by time. It'll just be an estimation of what the rate is based on that. Generally some sort of regression would provide a much better estimate of the rates at any given time, but this'll do for now.

For raw queries the duration is really just used to change the units on the derivative from per nanosecond to whatever duration they want.

For derivatives on the group by times, which are the ones in buckets, you actually want them to be consistent. i.e. use the first from each bucket or the last, or the max.

For example, if you have a measurement called `redis_key_count` and you write that in every `10s`. Then you do a query like this:

```sql
select non_negative_derivative(min(value), 5m) from redis_key_count
where time > now() - 24h
group by time(5m)
```

You wouldn't want to use the last from one bucket and the first from the next. It would give you a rate that is much smaller than what you'd expect for `5m` intervals.

Further, since counters are always monotonically increasing as long as it hasn't rolled over or Redis been reset, the `min` would functionally be the same as `first` just like `max` would be the same as `last`